In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

warnings.filterwarnings('ignore')
sns.set(style="whitegrid")

In [4]:
education_df = pd.read_csv('./states_all.csv')
education_df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

   ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  \
0  ...     57948.0     58025.0      41167.0      471564.0       196386.0   
1  ...      9748.0      8789.0       6714.0       79117.0        30847.0   
2  ...     55433.0     49081.0      37410.0      437127.0       175210.0   
3  ...     34632.0     36011.0      27651.0      281338.0       123113.0   
4  ...    418418.0    363296.0     270675.0     3286034.0      1372011.0   

   GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0      676174.0        208.327876        252.187522           207.963517   
1      112335.0               NaN               NaN                  NaN   
2      614881.0        215.253932        265.366278           206.212716   
3      405259.0        210.206028        256.312090           208.634458   
4     4717112.0        208.398961        260.892247           196.764414   

   AVG_READING_8_SCORE  
0                  NaN  
1           258.859712  
2           262.169895  
3           264.619665  
4                  NaN  

[5 rows x 25 columns]

In [5]:
fill_list = ["ENROLL", "TOTAL_REVENUE", "FEDERAL_REVENUE",
             "STATE_REVENUE", "LOCAL_REVENUE", "TOTAL_EXPENDITURE",
             "INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
             "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE", "GRADES_PK_G",
             "GRADES_KG_G", "GRADES_4_G", "GRADES_8_G", "GRADES_12_G", "GRADES_1_8_G",
             "GRADES_9_12_G", "GRADES_ALL_G"]

states = education_df["STATE"].unique()

for state in states:
    education_df.loc[education_df["STATE"] == state, fill_list] = education_df.loc[education_df["STATE"] == state, fill_list].interpolate()
education_df.dropna(inplace=True)

Ağırlıklı Ortalama

In [6]:
education_df["overall_score"] = (education_df["GRADES_4_G"]*((education_df["AVG_MATH_4_SCORE"] + education_df["AVG_READING_4_SCORE"])*0.5) + education_df["GRADES_8_G"]
                                 * ((education_df["AVG_MATH_8_SCORE"] + education_df["AVG_READING_8_SCORE"])*0.5))/(education_df["GRADES_4_G"] + education_df["GRADES_8_G"])

In [7]:
education_df.head()

PRIMARY_KEY       STATE  YEAR     ENROLL  TOTAL_REVENUE  \
204     1996_ALABAMA     ALABAMA  1996   735912.0      3365113.0   
206     1996_ARIZONA     ARIZONA  1996   764681.0      4143304.0   
207    1996_ARKANSAS    ARKANSAS  1996   452907.0      2183384.0   
208  1996_CALIFORNIA  CALIFORNIA  1996  5460484.0     31282981.0   
209    1996_COLORADO    COLORADO  1996   655679.0      3891203.0   

     FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
204         334858.0      2138717.0       891538.0          3572283.0   
206         378917.0      1798363.0      1966024.0          4391555.0   
207         192152.0      1296247.0       694985.0          2297381.0   
208        2603882.0     17061474.0     11617625.0         31561692.0   
209         194998.0      1664676.0      2031529.0          4075165.0   

     INSTRUCTION_EXPENDITURE  ...  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
204                1987018.0  ...     58305.0      42510.0      469370.0   
206                1860007.0  ...     60514.0      42041.0      511925.0   
207                1264275.0  ...     37064.0      27613.0      283883.0   
208               16880730.0  ...    406456.0     298669.0     3516572.0   
209                2052264.0  ...     52269.0      37179.0      423510.0   

     GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
204       207980.0      686983.0        211.646974        256.594863   
206       210289.0      726709.0        217.575940        267.874834   
207       131997.0      417860.0        215.846436        261.652745   
208      1525618.0     5115423.0        209.129083        262.772478   
209       185700.0      621730.0        225.805858        275.608214   

     AVG_READING_4_SCORE  AVG_READING_8_SCORE  overall_score  
204           210.920767           260.276441     235.106342  
206           206.529788           263.283638     238.321404  
207           208.837716           271.834590     240.773675  
208           201.912591           269.197451     234.719834  
209           221.653243           268.168154     247.750273  

[5 rows x 26 columns]

In [8]:
education_df["overall_score"] = (education_df["GRADES_4_G"]*((education_df["AVG_MATH_4_SCORE"] + education_df["AVG_READING_4_SCORE"])*0.5) + education_df["GRADES_8_G"]
                                 * ((education_df["AVG_MATH_8_SCORE"] + education_df["AVG_READING_8_SCORE"])*0.5))/(education_df["GRADES_4_G"] + education_df["GRADES_8_G"])

In [14]:
education_df[["overall_score", "TOTAL_EXPENDITURE", "INSTRUCTION_EXPENDITURE",
              "SUPPORT_SERVICES_EXPENDITURE", "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]].corr()

overall_score  TOTAL_EXPENDITURE  \
overall_score                      1.000000           0.074382   
TOTAL_EXPENDITURE                  0.074382           1.000000   
INSTRUCTION_EXPENDITURE            0.087626           0.990203   
SUPPORT_SERVICES_EXPENDITURE       0.079175           0.992861   
OTHER_EXPENDITURE                  0.011099           0.943844   
CAPITAL_OUTLAY_EXPENDITURE        -0.004472           0.924816   

                              INSTRUCTION_EXPENDITURE  \
overall_score                                0.087626   
TOTAL_EXPENDITURE                            0.990203   
INSTRUCTION_EXPENDITURE                      1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.974425   
OTHER_EXPENDITURE                            0.902323   
CAPITAL_OUTLAY_EXPENDITURE                   0.883155   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  \
overall_score                                     0.079175           0.011099   
TOTAL_EXPENDITURE                                 0.992861           0.943844   
INSTRUCTION_EXPENDITURE                           0.974425           0.902323   
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.952062   
OTHER_EXPENDITURE                                 0.952062           1.000000   
CAPITAL_OUTLAY_EXPENDITURE                        0.909381           0.917785   

                              CAPITAL_OUTLAY_EXPENDITURE  
overall_score                                  -0.004472  
TOTAL_EXPENDITURE                               0.924816  
INSTRUCTION_EXPENDITURE                         0.883155  
SUPPORT_SERVICES_EXPENDITURE                    0.909381  
OTHER_EXPENDITURE                               0.917785  
CAPITAL_OUTLAY_EXPENDITURE                      1.000000

Instruction_Expenditure ile overall_score arasındaki korelasyon en fazladır.

In [15]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

X = education_df[["INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
                  "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]]

X = StandardScaler().fit_transform(X)

sklearn_pca = PCA(n_components=1)
education_df["pca_1"] = sklearn_pca.fit_transform(X)

print('The percentage of total variance in the dataset explained by each','component from Sklearn PCA.\n',
      sklearn_pca.explained_variance_ratio_)

The percentage of total variance in the dataset explained by each component from Sklearn PCA.
 [0.94252701]


%94 den fazlası PCA ile açıklanır.

In [16]:
education_df[["overall_score", "pca_1", "TOTAL_EXPENDITURE", "INSTRUCTION_EXPENDITURE",
              "SUPPORT_SERVICES_EXPENDITURE", "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]].corr()

overall_score     pca_1  TOTAL_EXPENDITURE  \
overall_score                      1.000000  0.044988           0.074382   
pca_1                              0.044988  1.000000           0.992059   
TOTAL_EXPENDITURE                  0.074382  0.992059           1.000000   
INSTRUCTION_EXPENDITURE            0.087626  0.968492           0.990203   
SUPPORT_SERVICES_EXPENDITURE       0.079175  0.988085           0.992861   
OTHER_EXPENDITURE                  0.011099  0.971480           0.943844   
CAPITAL_OUTLAY_EXPENDITURE        -0.004472  0.955011           0.924816   

                              INSTRUCTION_EXPENDITURE  \
overall_score                                0.087626   
pca_1                                        0.968492   
TOTAL_EXPENDITURE                            0.990203   
INSTRUCTION_EXPENDITURE                      1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.974425   
OTHER_EXPENDITURE                            0.902323   
CAPITAL_OUTLAY_EXPENDITURE                   0.883155   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  \
overall_score                                     0.079175           0.011099   
pca_1                                             0.988085           0.971480   
TOTAL_EXPENDITURE                                 0.992861           0.943844   
INSTRUCTION_EXPENDITURE                           0.974425           0.902323   
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.952062   
OTHER_EXPENDITURE                                 0.952062           1.000000   
CAPITAL_OUTLAY_EXPENDITURE                        0.909381           0.917785   

                              CAPITAL_OUTLAY_EXPENDITURE  
overall_score                                  -0.004472  
pca_1                                           0.955011  
TOTAL_EXPENDITURE                               0.924816  
INSTRUCTION_EXPENDITURE                         0.883155  
SUPPORT_SERVICES_EXPENDITURE                    0.909381  
OTHER_EXPENDITURE                               0.917785  
CAPITAL_OUTLAY_EXPENDITURE                      1.000000

Overall_score ile pca_1 arasındaki korelasyon 0.044988 dir.

PCA, en iyi korelasyonun =<0.8 olduğu durumalarda çalışır. Ama harcama değerleri arasındaki korelasyon yüksektir. İlk temel değişkenleri tercih ederim.